### PR TITLE
Move SFPU based activation to packer thread in BMM kernels

### PR DESCRIPTION
**Pull request by iwroszTT**
### Summary
Issue: https://github.com/tenstorrent/tt-metal/issues/37696
Matmul kernels with fused activation currently serialize matrix multiply with the calculation of the activation. To increase performance we want to overlay matmul with activation by moving the SFPU-based activation from MATH to the PACK thread.

Performance comparison
- Synthetic matmul (`MatmulMultiCoreReuseMultiCast1DProgramConfig`) test M=2048 N=512 and variable K, see the attached script
[simple_matmul_test.py](https://github.com/user-attachments/files/27207098/simple_matmul_test.py)

  - Table shows relative device time of mm with activation done on PACK thread vs current (lower is better)
  - Note 1: SILU algorithm is more compute heavy than RELU6
  - Note 2: activation is done per output element (MxN) so by changing K we can test different ratio of matmul work vs activation (activation work is same of any K, but matmul work scales linearly)
  - Observation: The measured speedup pattern reflects the expected behavior (more speedup for SILU vs RELU6, more speedup for smaller K)
  
  | K    | SILU  | RELU6 |                                                               
  |------|-------|-------|                                                             
  | 32   | 0.809 | 0.989 |                                                               
  | 64   | 0.815 | 0.985 |                                                               
  | 128  | 0.831 | 0.984 |
  | 256  | 0.860 | 0.988 |                                                               
  | 512  | 0.890 | 0.999 |                                                               
  | 1024 | 0.925 | 0.997 |

- Single Card Model Perf Tests
  - PR Run: https://github.com/tenstorrent/tt-metal/actions/runs/24989450636
  - Main Run: https://github.com/tenstorrent/tt-metal/actions/runs/24982944336
  - PR: https://github.com/tenstorrent/tt-metal/pull/43067
  
  **Metric: Inference Time Change (%)** — perf impact within test noise        
                                                            
  | Model | Wormhole B0 | Blackhole |                                                     
  |-------|-------------|-----------|                       
  | **CNN/Vision Models** | | |
  | StableDiffusion | +0.18% | +0.01% |
  | UNet (single) | 0.00% | — |                                                           
  | UNet (multi-device) | 0.00% | — |
  | **Language Models** | | |                                                             
  | BERT11 | **-1.57%**  | — |                            
  | DistilBERT | +0.31% | -0.65% |                                                        
  | **Classification** | | |
  | ResNet50 (batch 16) | 0.00% | 0.00% |                                                 
  | ResNet50 (batch 32) | — | 0.00% |                                                     
  | VoVNet | **-3.12%** | — |
  | | | |                                                                                                                                      

### Notes for reviewers
 - The kernel code uses semaphore calls to synchronize MATH and PACK accesses to Dest
 - The get_defines() infrastructure for selecting LLK SFPU functions is replaced with kernel side compile time dispatch structures.
 -  New ckernel functions are added (e.g., silu_tile_pack), which apply the PACK() wrapper instead of MATH()
 - Tests added for matmul operations with fused activation

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:iwrosz/sfpu-pack-thread)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:iwrosz/sfpu-pack-thread)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:iwrosz/sfpu-pack-thread)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:iwrosz/sfpu-pack-thread)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:iwrosz/sfpu-pack-thread)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:iwrosz/sfpu-pack-thread)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:iwrosz/sfpu-pack-thread)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:iwrosz/sfpu-pack-thread)

### CODE REVIEW

**Comment by iwroszTT** [comment]
@copilot Apply the fixes suggested in github-code-quality bot's review comments. Apply all fixes in a single commit.

**Review by ncvetkovicTT — APPROVED** [review]

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py [hunk 1]
```diff
@@ -0,0 +1,1115 @@
...
+        pt_out.float(),
+        tt_out,
+        atol=atol,
+        rtol=rtol,
+        frobenius_threshold=0.01 * K,
+        pcc_threshold=0.98,
+        check_ulp=False,
+    )
+
+    logger.info(f"✓ Matmul with {activation} activation passed")
```
Comment (bbradelTT) [inline tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py]: Unnecessary log bloat. Please remove.

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py [hunk 2]
```diff
@@ -0,0 +1,1115 @@
...
+        pt_out.float(),
+        tt_out,
+        atol=0.05 * K,
+        rtol=15.0 * K,
+        frobenius_threshold=0.02 * K,
+        pcc_threshold=0.95,
+        check_ulp=False,
+    )
+
+    logger.info(f"✓ Matmul with custom {activation.op_type} parameters passed")
```
Comment (bbradelTT) [inline tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py]: Please remove.

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py [hunk 3]
```diff
@@ -0,0 +1,1115 @@
...
+        pt_out.float(),
+        tt_out,
+        atol=0.03 * K,
+        rtol=12.0 * K,
+        frobenius_threshold=0.01 * K,
+        pcc_threshold=0.97,
+        check_ulp=False,
+    )
+
+    logger.info(f"✓ {config_type} config with {activation} activation passed")
```
Comment (bbradelTT) [inline tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py]: Please remove.

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_large_block_zm_fused_bias_activation_gathered.cpp [hunk 4]
```diff
@@ -373,19 +382,32 @@ void kernel_main() {
...
+                        // Flip destination register offset for PACKER access
+                        PACK(TT_SETC16(
+                            DEST_TARGET_REG_CFG_MATH_Offset_ADDR32, ckernel::packer::get_packer_dest_offset()));
+
+                        for (uint32_t i = 0; i < out_subblock_num_tiles; i++) {
+                            ActivationApplyHelper<activation_type, activation_param0, activation_param1>::apply(i);
+                        }
+
+                        // Wait for SFPU completion before packing
+                        PACK(TTI_STALLWAIT(p_stall::STALL_PACK, p_stall::WAIT_SFPU));
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_large_block_zm_fused_bias_activation_gathered.cpp]: Can this be moved into a common function?

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py [hunk 5]
```diff
@@ -0,0 +1,1115 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent USA, Inc.
```
Comment (bbradelTT) [inline tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py]: 2026?

**Review by bbradelTT — COMMENTED** [review]
Please comment on or address AI comments.

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py [hunk 6]
```diff
@@ -0,0 +1,1115 @@
...
+            "gelu": ttnn.UnaryWithParam(ttnn.UnaryOpType.GELU),
+            "tanh": ttnn.UnaryWithParam(ttnn.UnaryOpType.TANH),
+            "sigmoid": ttnn.UnaryWithParam(ttnn.UnaryOpType.SIGMOID),
+            "hardsigmoid": ttnn.UnaryWithParam(ttnn.UnaryOpType.HARDSIGMOID),
+            "hardtanh": ttnn.UnaryWithParam(ttnn.UnaryOpType.HARDTANH),
+            "selu": ttnn.UnaryWithParam(ttnn.UnaryOpType.SELU),
+            "softplus": ttnn.UnaryWithParam(ttnn.UnaryOpType.SOFTPLUS),
+            "mish": ttnn.UnaryWithParam(ttnn.UnaryOpType.MISH),
+        }
+        fused_activation = activation_map.get(activation, None)
```
[the remark this answers is not in the document]
Comment (iwroszTT) [reply tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py:190]: mish activation is included in the string to UnaryOpType mapping for future use, it's not included in test parameters, hence it's not tested

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py (hunk 1 above)
Comment (iwroszTT) [reply tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py]: Removed

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_large_block_zm_fused_bias_activation_gathered.cpp (hunk 4 above)
Comment (iwroszTT) [reply ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_large_block_zm_fused_bias_activation_gathered.cpp]: Created common function

**Review by bbradelTT — APPROVED** [review]

File: tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py [hunk 7]
```diff
@@ -0,0 +1,1111 @@
...
+    has_bias,
+    activation,
+    grid_size,
+    in0_dtype,
+    in1_dtype,
+    out_dtype,
+    function_level_defaults,
+):
+    """Test DRAM sharded matmul with combinations of bias and activation."""
+    logger.info(f"Testing DRAM sharded matmul: M={M}, K={K}, N={N}, bias={has_bias}, activation={activation}")
```
Comment (bbradelTT) [inline tests/ttnn/nightly/unit_tests/operations/matmul/test_matmul_activations.py]: This is all part of the pytest parameters. It'd be good to remove these except for maybe the lower pcc one.

**Review by dpopovTT — APPROVED** [review]

File: tt_metal/hw/inc/api/compute/eltwise_unary/relu.h [hunk 8]
```diff
@@ -50,12 +50,16 @@ ALWI void relu_tile(uint32_t idst) { MATH(SFPU_UNARY_ONE_PARAM_KERNEL_FN_FLOAT(_
 ALWI void relu_max_tile(uint32_t idst, uint32_t param0) {
     MATH(SFPU_UNARY_ONE_PARAM_KERNEL_FN_FLOAT(_relu_max_, RC, APPROX, idst, param0));
 }
+ALWI void relu_max_tile_pack(uint32_t idst, uint32_t param0) {
```
Comment (fvranicTT) [inline tt_metal/hw/inc/api/compute/eltwise_unary/relu.h:53]: Technically, this one should only be present on pack, so it could be wrapped into a separate ifdef, so the or from above should he refactored.

Comment (iwroszTT) [reply tt_metal/hw/inc/api/compute/eltwise_unary/relu.h:53]: If I understand the comment correctly, the wrapped function would look like this:
```
#ifdef TRISCPACK
ALWI void relu_max_tile_pack(uint32_t idst, uint32_t param0) {
    // perhaps without the PACK wrapper
    PACK(SFPU_UNARY_ONE_PARAM_KERNEL_FN_FLOAT(_relu_max_, RC, APPROX, idst, param0));
}
#endif
```

In these changes, I tried to follow the existing pattern, where the compute api functions are visible on all threads, although their implementation is limited to a certain thread, typically MATH. So that we have:
```
ALWI void relu_max_tile(uint32_t idst, uint32_t param0) {
    MATH(SFPU_UNARY_ONE_PARAM_KERNEL_FN_FLOAT(_relu_max_, RC, APPROX, idst, param0));
}
```
and not:
```
#ifdef TRISCMATH
ALWI void relu_max_tile(uint32_t idst, uint32_t param0) {
    MATH(SFPU_UNARY_ONE_PARAM_KERNEL_FN_FLOAT(_relu_max_, RC, APPROX, idst, param0));
}
#endif
```

Comment (fvranicTT) [reply tt_metal/hw/inc/api/compute/eltwise_unary/relu.h:53]: Sure, sorry for trying to invent something new. I was just trying to limit the exposure to the minimal number of threads that need it. I now realize there is another pattern in the codebase. I'll sync with the team wrt that. I don't want to block on that. Please continue the way you are doing it now.

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp [hunk 9]
```diff
@@ -902,7 +905,8 @@ UnaryWithParam string_to_unary_with_param(const std::string& name) {
         return UnaryWithParam(UnaryOpType::XIELU, {0.8f, 0.8f});  // alpha_p=0.8, alpha_n=0.8
     }
     if (name == "selu") {
-        return UnaryWithParam(UnaryOpType::SELU);
+        return UnaryWithParam(
+            UnaryOpType::SELU, {1.67326f, 1.05070f});  // alpha=1.67326, lambda=1.05070 (default values)
```
Comment (mateusznowakTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp]: I think we need few digits more to exhaust fp32 accuracy

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_fused_activation.hpp [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+            ACT == KernelActivation::SELU || ACT == KernelActivation::SOFTPLUS,
+        "Unsupported KernelActivation type for fused activation init");
+
+    FORCE_INLINE static void init() {
+        if constexpr (ACT == KernelActivation::SILU) {
+            silu_tile_init_pack();
+        } else if constexpr (ACT == KernelActivation::TANH) {
+            // PARAM0: 0 = accurate, non-zero = fast
+            if constexpr (PARAM0 != 0) {
+                tanh_tile_init_pack<true>();
```
Comment (fvranicTT) [inline ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_fused_activation.hpp]: You can collapse this into:
```cpp
tanh_tile_init_pack<(PARAM0 != 0)>();
```
Others too.

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp (hunk 9 above)
Comment (iwroszTT) [reply ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp]: Changed to fp32 representation of the values given here: https://docs.pytorch.org/docs/2.11/generated/torch.nn.modules.activation.SELU.html

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_fused_activation.hpp (hunk 10 above)
Comment (iwroszTT) [reply ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_fused_activation.hpp]: done

**Review by mateusznowakTT — APPROVED** [review]

**Review by abhullar-tt — APPROVED** [review]
Genfiles changes lgtm
